drivers/misc/block_cache.c

```diff
@@ -0,0 +1,921 @@
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#if defined(CONFIG_BLOCK_CACHE_DEBUG)
+# define bcinfo _info
+#else
+# define bcinfo _none
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+enum block_flags
+{
+  BLOCK_FREE  = 0x1,                  /* The window is marked as free */
+  BLOCK_DIRTY = 0x2,                  /* The window contains dirty blocks*/
+};
+
+struct cache_window_s
+{
+  blkcnt_t          start;            /* The initial block index */
+  enum block_flags  flags;            /* Any flags for this window */
+  uint8_t*          buffer;           /* The underlying block buffer */
+};
+
+
+struct block_cache_dev_s
+{
+  FAR struct inode *inode;                /* Contained block device*/
+  struct geometry geo;                    /* Contained device geometry */
+  size_t geo_multiple;                    /* External block multiple */
+  struct cache_window_s* cache;           /* Array of block caches*/
+  off_t                 last_index;       /* The last index evicted*/
+  uint16_t              refs;             /* Number of references */
+  size_t                cache_width;      /* The width of a cache*/
+  size_t                cache_count;      /* The number of caches */
+  size_t                cache_used_count; /* Number of used caches*/
+  size_t                aligned_mask;     /* The mask for block alignment*/
+  mutex_t               lock;             /* Lock */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+/* Block device handles */
+
+static int     block_cache_open(FAR struct inode *inode);
+static int     block_cache_close(FAR struct inode *inode);
+static ssize_t block_cache_read(FAR struct inode *inode, FAR unsigned char *buffer,
+                 blkcnt_t start_sector, unsigned int nsectors);
+static ssize_t block_cache_write(FAR struct inode *inode,
+                 FAR const unsigned char *buffer, blkcnt_t start_block,
+                 unsigned int nblocks);
+static int     block_cache_geometry(FAR struct inode *inode,
+                 FAR struct geometry *geometry);
+static int     block_cache_ioctl(FAR struct inode *inode, int cmd,
+                 unsigned long arg);
+
+
+/* Interface to the underlying block device*/
+
+static ssize_t block_cache_reload(struct block_cache_dev_s*, FAR uint8_t *buffer,
+                 off_t startblock, size_t nblocks);
+static ssize_t block_cache_sync(struct block_cache_dev_s*, FAR const uint8_t *buffer,
+                 off_t startblock, size_t nblocks);
+
+/* Internal use  */
+
+int block_cache_write_internal(struct block_cache_dev_s* dev, const unsigned char *buffer,
```

Comment:
could we merge the new cache algo to:
https://github.com/apache/nuttx/blob/master/drivers/misc/rwbuffer.c
so, many driver(e.g. bch and ftl) can get the benefit immediately